core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java

```diff
@@ -34,6 +48,120 @@
+      protected final Boolean addValidResponse(Address sender, ValidResponse response) {
+         return (Boolean) response.getResponseValue();
+      }
+
+      abstract Boolean addBooleanResponse(Address sender, Boolean response);
+   }
+
+   private static class ScatteredTouchResponseCollector extends AbstractTouchResponseCollector {
+
+      private static final ScatteredTouchResponseCollector INSTANCE = new ScatteredTouchResponseCollector();
+
+      @Override
+      public Boolean finish() {
+         // No other node was touched
+         return Boolean.FALSE;
+      }
+
+      @Override
+      protected Boolean addBooleanResponse(Address sender, Boolean response) {
+         if (response == Boolean.TRUE) {
+            // Return early if any node touched the value - as SCATTERED only exists on a single backup!
+            // TODO: what if the read was when one of the backups or primary died?
+            return Boolean.TRUE;
+         }
+         return null;
+      }
+   }
+
+   private static class TouchResponseCollector extends AbstractTouchResponseCollector {
+
+      private static final TouchResponseCollector INSTANCE = new TouchResponseCollector();
+
+      @Override
+      public Boolean finish() {
+         // If all were touched, then the value isn't expired
+         return Boolean.TRUE;
+      }
+
+      @Override
+      protected Boolean addBooleanResponse(Address sender, Boolean response) {
+         if (response == Boolean.FALSE) {
+            // Return early if any value wasn't touched!
+            return Boolean.FALSE;
+         }
+         return null;
+      }
+   }
+
+   @Override
+   public Object visitTouchCommand(InvocationContext ctx, TouchCommand command) throws Throwable {
+      if (isLocalModeForced(command)) {
+         return invokeNext(ctx, command);
+      }
+      LocalizedCacheTopology cacheTopology = checkTopologyId(command);
+      DistributionInfo info = cacheTopology.getSegmentDistribution(command.getSegment());
+
+      if (info.isPrimary()) {
+         boolean isScattered = cacheConfiguration.clustering().cacheMode().isScattered();
+         // Scattered any node could be a backup, so we have to touch all members
+         List<Address> owners = isScattered ? cacheTopology.getActualMembers() : info.readOwners();
+         AbstractTouchResponseCollector collector = isScattered ? ScatteredTouchResponseCollector.INSTANCE :
+               TouchResponseCollector.INSTANCE;
+         CompletionStage<Boolean> remoteInvocation = rpcManager.invokeCommand(owners, command, collector,
+               rpcManager.getSyncRpcOptions());
+         return invokeNextThenApply(ctx, command, (rCtx, rCommand, rValue) -> {
+            Boolean touchedLocally = (Boolean) rValue;
+            Boolean collectedResponse = collector.addBooleanResponse(null, touchedLocally);
+            if (collectedResponse != null) {
+               return asyncValue(remoteInvocation.thenApply(ignore -> touchedLocally));
```

Comment:
why do you need to wait for the remote operation? can't you return `touchedLocally` directly?

Reply:
It needs to be sync to maintain the cluster max idle guarantees that if a node goes down after a touch that it won't expire early.